### PR TITLE
Fixes #10369 - added support for scenarios

### PR DESCRIPTION
**Pull request by mbacovsky**
This change adds support for scenarios. Most of the logic is part of Kafo. In the installer we have to implement:
- update binary installer to point to scenarios dir
- convert configuration to scenarios and change location
- updates in man page
- added new hook responsible for migrating custom data from existing old config to scenario + answers

Scratch build for testing: https://koji.katello.org/koji/taskinfo?taskID=407578

### CODE REVIEW

File: config/foreman.yaml [hunk 1]
```diff
@@ -27,6 +27,7 @@
 
 ## Kafo tuning, customization of core functionality
 :name: Foreman
+:descritpion: Default instalation of Foreman
```
Comment (mmoll) [inline config/foreman.yaml]: typo there? :musical_keyboard:

Comment (mbacovsky) [inline config/foreman.yaml]: Thanks, fixed. :notes:

File: config/foreman.yaml [hunk 2]
```diff
@@ -27,6 +27,7 @@
 
 ## Kafo tuning, customization of core functionality
 :name: Foreman
+:descritpion: Default installation of Foreman
```
Comment (mmoll) [inline config/foreman.yaml]: unfortunately that's not all... "description"

**Comment by ares** [comment]
depends on https://github.com/theforeman/kafo/pull/126

**Comment by ares** [comment]
tested with the linked PR and works for me well, I didn't do the full review though, since I used is only as kafo tester

File: bin/foreman-installer [hunk 3]
```diff
@@ -2,17 +2,8 @@
...
-config_filename = File.basename($0) + ".yaml"
-
-# where to find answer file
-if File.exist?('config/' + config_filename)
-  CONFIG_FILE = 'config/' + config_filename
-else
-  CONFIG_FILE = '/etc/foreman/' + config_filename
-end
+# where to find scenarios
+CONFIG_DIR = '/etc/foreman/installer-scenarios.d/'
```
Comment (ehelms) [inline bin/foreman-installer]: The previous logic allowed running of the installer from a git checkout which is nice when testing changes to modules or the installer logic itself. Could we keep that?

Comment (mbacovsky) [inline bin/foreman-installer]: Ok, would you prefer `config/` or `/config/installer-scenarios.d/`?

File: config/foreman.yaml [hunk 4]
```diff
@@ -3,14 +3,14 @@
 
 ## Installer configuration
 # Path to answer file
-:answer_file: ./config/answers.yaml
+:answer_file: ./config/foreman-answers.yaml
 :installer_dir: .
 # Uncomment if you want to load puppet modules from a specific path, $pwd/modules is used by default
-:modules_dir: ./_build/modules
+:module_dirs: ./_build/modules
```
Comment (ehelms) [inline config/foreman.yaml:9]: I know this is here from before, but if this is reduced to './modules', then for running from git I can easily do:

git clone https://github.com/theforeman/foreman-installer.git
gem install kafo or RPM install
cd foreman-installer
./bin/foreman-installer -v

**Comment by ehelms** [comment]
I ran this on a fresh EL7 and a 1.10 machine upgrading to nightly and installing the scratch RPM. :+1:

File: config/foreman.yaml (hunk 4 above)
Comment (ehelms) [inline config/foreman.yaml:9]: I will retract this comment

**Comment by mbacovsky** [comment]
Is there anything I can do to help this PR to get merged? 
It would help us immensely if we can get this to nightlies. Testing of katello conversion to scenarios is relying on these changes.

**Comment by ohadlevy** [comment]
anyone actively reviewing this change?

File: hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    # copy over config and answers
+    scenario.migrate_configuration(legacy_config)
+    scenario.store(legacy_config.answers)
+
+    # link last used scenario
+    kafo.class.scenario_manager.link_last_scenario(scenario.config_file)
+    logger.info("Scenario #{scenario.config_file} was detected to be the last used scenario")
+
+    kafo.request_config_reload
+    logger.info("Configuration #{legacy_config_file} was successfuly migrated. Installer configuration reload was requested.")
```
Comment (domcleal) [inline hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb]: "successfully"

File: hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  # rename legacy config and answer file
+  File.rename(legacy_config.answer_file, legacy_config.answer_file + '.backup')
+  File.rename(legacy_config.config_file, legacy_config.config_file + '.backup')
+  logger.info("Legacy configuration #{legacy_config_file} was removed.")
```
Comment (domcleal) [inline hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb]: It's not removed, it's migrated and backed up?

File: man/foreman-installer.8.asciidoc [hunk 7]
```diff
@@ -12,6 +12,8 @@ SYNOPSIS
 --------
 foreman-installer
 
+foreman-installer --scenario foreman
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc:17]: The `--list-scenarios` call may need to be shown here too.

File: man/foreman-installer.8.asciidoc [hunk 8]
```diff
@@ -141,7 +143,29 @@ Enable 'puppet' puppet module (default: true)
...
 Drops the Foreman database, then runs the installer to recreate and reset it.
-This is a PERMANENT action and will destroy all data stored in Foreman.
+This is a PERMANENT action and will destroy all data stored in Foreman
+
+--detailed-exitcodes::
+Provide transaction information via exit codes, see puppet-agent(8) for
+full details. (default: false)
+
+-p, --profile::
+Run puppet in profile mode? (default: false)
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc]: Capitalise Puppet.

File: man/foreman-installer.8.asciidoc [hunk 9]
```diff
@@ -141,7 +143,29 @@ Enable 'puppet' puppet module (default: true)
 
 --reset-foreman-db::
 Drops the Foreman database, then runs the installer to recreate and reset it.
-This is a PERMANENT action and will destroy all data stored in Foreman.
+This is a PERMANENT action and will destroy all data stored in Foreman
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc]: Keep this, as it's a separate sentence?

Comment (domcleal) [inline man/foreman-installer.8.asciidoc]: Keep this, as it's a separate sentence?

File: man/foreman-installer.8.asciidoc [hunk 10]
```diff
@@ -155,13 +179,17 @@ The following parameters are available as installer arguments:
...
 --------
 
+List available scenarios
+
+  foreman-installer --list-scenarios
+
 Install with default options
 
-  foreman-installer
+  foreman-installer --scenario foreman
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc:196]: Should be a separate example?

File: man/foreman-installer.8.asciidoc [hunk 11]
```diff
@@ -175,9 +203,9 @@ Install TFTP, DHCP, DNS with provided options
 FILES
 -----
 
-`/etc/foreman/foreman-installer.yaml`::
-Foreman-installer is based on the 'kafo' project. Contains configuration for
+`/etc/foreman/installer-scenarios.d/foreman.yaml`::
+Foreman-installer is based on the 'kafo' project. Contains configurations (a.k.a. scenarios) for
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc]: I think the original is clearer, it's referring to a single configuration.

File: hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb (hunk 6 above)
Comment (mbacovsky) [inline hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb]: Yes, it seemed better to keep the original file for eventual future investigation.
Would you prefer differently?

Comment (domcleal) [inline hooks/pre_migrations/01-migrate_legacy_foreman_installer_config.rb]: That's fine, but this log line should probably indicate that.

File: man/foreman-installer.8.asciidoc [hunk 12]
```diff
@@ -86,7 +90,7 @@ parameters, and can be run by executing:
 
 The answers file describes the classes that will be applied to the host to
 install Foreman, along with their parameters. The foreman-installer package
-stores it at `/etc/foreman/foreman-installer.yaml`. By default, the all-in-one
+stores it at `/etc/foreman/installer-scenarios.d/foreman-answers.yaml`. By default, the all-in-one
```
Comment (domcleal) [inline man/foreman-installer.8.asciidoc]: line wrapping should match

**Comment by domcleal** [comment]
Merged as 401c0d9, thanks @mbacovsky.

**Pull request closed without merging**